js-bmfbp/prototype.js

```diff
@@ -272,91 +368,197 @@ function bmfbp(schematic) {
     };
   }
 
-  function send(originatingPartIndex, pinIndex, packet) {
-    outQueue[originatingPartIndex].push(OutEvent(packet, originatingPartIndex, pinIndex));
-  }
-
-  function pushToInQueue(wireNumber, outEvent) {
-    const receivers = wireToReceivers[wireNumber];
-    var i, l;
-    var receiver;
-
-    for (i = 0, l = receivers.length; i < l; i++) {
-      receiver = receivers[i];
-      readyParts.push(receiver.part);
-      inQueue[receiver.part].push({
-        pin: receiver.pin,
-        packet: outEvent.packet
-      });
+  function initComposite(schematic) {
+    // Maps Part index in `schematic.parts` to Part object
+    const parts = new Array(schematic.parts.length);
+    // The main entry points of the Parts, indexed like `parts` above
+    const partMains = new Array(schematic.parts.length);
+    // Maps wire number to the Parts that are on the receiving end of the wire
+    const wireToReceivers = new Array(schematic.wireCount);
+    // The input and output queues of the Parts, indexed by the Part index.
+    // There is exactly one input queue and one output queue for each Part.
+    const inQueue = new Array(schematic.parts.length);
+    // The output queue is an array of arrays, first indexed by Part index,
+    // then by the Part's pin index.
+    const outQueue = new Array(schematic.parts.length);
+    // Indexes of the Parts that can be activated
+    const readyParts = [];
+    const wireToSource = schematic.sources;
+    // We need to transform the `sinks` in schematic, which maps Sink number
+    // to Wire number, to an array that maps Wire number to Sink number for
+    // efficiency at run-time.
+    const wireToSinks = new Array(schematic.wireCount);
+    // This is the `send()` subroutine to send packets via the sinks of this
+    // composite. We don't have access to this until the parent Composite
+    // calls the main function of this Composite.
+    var compositeSend = null;
+    // We need a way to know when to flush the Sinks.
+    var flushSinks = false;
```

Comment:
@guitarvydas I forgot to point this out. I ended up using a flag called `flushSinks` for each composite to keep track of that we have packets sent through the sinks. I found that I needed it because dispatch only checks whether there is any ready Parts inside the Composite. Is there another way to do this without using a flag?